tests/integration_tests/test_upgrade.py

```diff
@@ -0,0 +1,71 @@
+import logging
+import pytest
+from pathlib import Path
+
+from tests.integration_tests.clouds import ImageSpecification, IntegrationCloud
+from tests.integration_tests.conftest import (
+    get_validated_source,
+    session_start_time,
+)
+
+log = logging.getLogger('integration_testing')
+
+
+def _output_to_compare(instance, file_path, netcfg_path):
+    commands = [
+        'hostname',
+        'dpkg-query --show cloud-init',
+        'cat /run/cloud-init/result.json',
+        'grep Trace /var/log/cloud-init.log',
+        'systemd-analyze',
+        'systemd-analyze blame',
+        'cloud-init analyze show',
+        'cloud-init analyze blame',
+        'cat {}'.format(netcfg_path),
+        'cloud-id'
+    ]
+    with file_path.open('w') as f:
+        for command in commands:
+            f.write('===== {} ====='.format(command) + '\n')
+            f.write(instance.execute(command) + '\n')
+
+
+@pytest.mark.sru_2020_11
+def test_upgrade(session_cloud: IntegrationCloud):
+    source = get_validated_source()
+    if not source.installs_new_version():
+        pytest.skip("Current install method not supported for this test")
+        return  # type checking doesn't understand that skip raises
+
+    launch_kwargs = {
+        'name': 'integration-upgrade-test',
+        'image_id': session_cloud._get_initial_image(),
+        'wait': True,
+    }
+
+    image = ImageSpecification.from_os_image()
+
+    # Get the paths to write test logs
+    output_dir = Path(session_cloud.settings.LOCAL_LOG_PATH)
+    output_dir.mkdir(parents=True, exist_ok=True)
+    base_filename = 'test_upgrade_{os}_{{stage}}_{time}.log'.format(
+        os=image.release,
+        time=session_start_time,
+    )
+    before_path = output_dir / base_filename.format(stage='before')
+    after_path = output_dir / base_filename.format(stage='after')
+
+    # Get the network cfg file
+    netcfg_path = '/dev/null'
+    if image.os == 'ubuntu':
+        netcfg_path = '/etc/netplan/50-cloud-init.yaml'
+        if image.release == 'xenial':
+            netcfg_path = '/etc/network/interfaces.d/50-cloud-init.cfg'
+
+    with session_cloud.launch(launch_kwargs=launch_kwargs) as instance:
+        _output_to_compare(instance, before_path, netcfg_path)
+        instance.install_new_cloud_init(source, take_snapshot=False)
+        instance.instance.restart()
```

Comment:
I think we need to fail on seeing tracebacks after reset.  Also I think we need to check cloud-init init and no Tracebacks as we do in [ec2 manual sru test](https://github.com/cloud-init/ubuntu-sru/blob/master/sru-scripts/manual/ec2-sru#L117-L118)


I also think we should try to inject some userdata that exercises parts of cloud-init on the upgrade test to ensure that various pieces are working

We could use and validate [sample jinja template](https://github.com/cloud-init/ubuntu-sru/blob/master/sru-scripts/sru-vars.template#L51-L55) as we did in manual test runs to assert that hostname returned actually contains the right cloud-id from instance data

So if we can use 

```suggestion
        instance.instance.execute('cloud-init init')
        tracebacks = instance.instance.execute('grep Traceback /var/log/cloud-init.log')
        assert tracebacks == ''
        instance.restart(raise_on_cloudinit_failure=True)   # Make sure we fail this test if cloud-init fails on next clean boot
```

Reply:
> I think we need to fail on seeing tracebacks after reset. Also I think we need to check cloud-init init and no Tracebacks as we do in ec2 manual sru test

Would `instance.restart(raise_on_cloudinit_failure=True)` accomplish this? Why would `cloud-init init` need to run again if it has already run on boot?

Reply:
@TheRealFalcon, some individuals who use cloud-init on a running system and don't want a reboot attempt to upgrade cloud-init and re start cloud-init services via `systemctl` this particular use-case, while not recommended, has raised a number of upgrade issues in the past with cloud-init and pickled DS cache handling across that upgrade path. So it's helpful for us to spot check this as it can be valuable in ascertaining if some of our pickling upgrade paths have broken across re-constitution of that cached datasource.  If we do this only after reboot, some platforms invalidate their datasource cache forcing a clean run instead of a dirty run of cloud-init. A fully clean run won't raise such pickle deserializing errors.